File: OneSWave/FullSpace/getVolNNSWave.m
function [IFull] = getVolNNSWave(rVec,zCav,vP,f,thetaW,phiW,wC)
%this function computes the volume contribution of NN from all nodes
% defined by rVec
% first get the amplitudes at all node points
lambda = vP/f;
k = 2*pi/lambda;
kVec = [k*sind(thetaW)*cosd(phiW);k*sind(thetaW)*sind(phiW);k*cosd(thetaW)];
%rVec = [x',y*ones(length(x),1),z*ones(length(x),1)];
i = sqrt(-1);
kr = rVec*kVec;
ampOut = exp(-i*(kr));
if(wC=='SH')
    kUnitVec = [-sind(phiW);cosd(phiW);0];
end
if(wC=='SV')
    kUnitVec = [cosd(thetaW)*cosd(phiW);cosd(thetaW)*sind(phiW);-sind(thetaW)];
end
%kUnitVec = kVec./(sqrt(kVec(1,1)^2+kVec(2,1)^2+kVec(3,1)^2));
ampOut = [ampOut*kUnitVec(1,1),ampOut*kUnitVec(2,1),ampOut*kUnitVec(3,1)];

% update the rVec for the current calculation
rVec(:,3) = rVec(:,3)-zCav; % (r-r0) vector
rDist = rVec.^2;
rDist = sum(rDist,2);
rDist = sqrt(rDist);
rCapX = rVec(:,1)./rDist;
rCapY = rVec(:,2)./rDist;
rCapZ = rVec(:,3)./rDist;
rDist3 = rDist.^3;

I1X = ampOut(:,1)./(rDist3);
I1Y = ampOut(:,2)./(rDist3);
I1Z = ampOut(:,3)./(rDist3);

% now get I2
I2 = (rCapX).*ampOut(:,1)+(rCapY).*ampOut(:,2)+(rCapZ).*ampOut(:,3);
I2 = 3*[I2.*(rCapX),I2.*(rCapY),I2.*(rCapZ)];
I2(:,1) = I2(:,1)./(rDist3);
I2(:,2) = I2(:,2)./(rDist3);
I2(:,3) = I2(:,3)./(rDist3);

IFull = [I1X-I2(:,1),I1Y-I2(:,2),I1Z-I2(:,3)];

end

File: OneSWave/FullSpace/getVolNNSWave.py
def getVolNNSwave(r_vec, z_cav, v_s, f, theta_w, phi_w, wC):
    
    theta_r = np.deg2rad(theta_w)
    phi_r = np.deg2rad(phi_w)

    k = 2 * np.pi * f / v_s
    k_vec = np.array([
        np.sin(theta_r) * np.cos(phi_r),
        np.sin(theta_r) * np.sin(phi_r),
        np.cos(theta_r)
    ]) * k

    kr = r_vec @ k_vec
    amp = np.exp(-1j * kr)

    if wC.upper() == 'SH':
        k_unit = np.array([-np.sin(phi_r), np.cos(phi_r), 0.0])
    elif wC.upper() == 'SV':
        k_unit = np.array([
            np.cos(theta_r) * np.cos(phi_r),
            np.cos(theta_r) * np.sin(phi_r),
            -np.sin(theta_r)
        ])
    else:
        raise ValueError("wC must be 'SH' or 'SV'")

    amp_vec = amp[:, None] * k_unit

    r = r_vec.copy()
    r[:, 2] -= z_cav
    r_dist2 = np.sum(r**2, axis=1)
    r_dist = np.sqrt(r_dist2)
    inv_r3 = 1.0 / (r_dist2 * r_dist)

    r_cap = r / r_dist[:, None]

    I1 = amp_vec * inv_r3[:, None]
    I2_dot = np.einsum('ij,ij->i', r_cap, amp_vec)
    I2 = 3 * I2_dot[:, None] * r_cap * inv_r3[:, None]

    return I1 - I2
